Source/Core/Resource.js

```diff
@@ -0,0 +1,183 @@
+define([
+    './combine',
+    './defaultValue',
+    './defined',
+    './defineProperties',
+    './joinUrls',
+    './objectToQuery',
+    './queryToObject',
+    './Check',
+    '../ThirdParty/Uri'
+], function(combine,
+            defaultValue,
+            defined,
+            defineProperties,
+            joinUrls,
+            objectToQuery,
+            queryToObject,
+            Check,
+            Uri) {
+    'use strict';
+
+    /**
+     * @param {Object} options An object with the following properties
+     * @param {String} options.url
+     * @param {Object} [options.queryParameters]
+     * @param {Object} [options.headers]
+     * @param {Request} [options.request]
+     * @param {String} [options.method='GET']
+     * @param {Object} [options.data]
+     * @param {String} [options.overrideMimeType]
+     * @param {DefaultProxy} [options.proxy]
+     * @param {Boolean} [options.allowCrossOrigin]
+     *
+     * @constructor
+     */
+    function Resource(options) {
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+
+        //>>includeStart('debug', pragmas.debug);
+        Check.defined('options.url', options.url);
+        //>>includeEnd('debug');
+
+        this._url = '';
+        this._queryParameters = defaultValue(options.queryParameters, {});
+        this._hash = defaultValue(options.hash, '');
+
+        this.url = options.url;
+
+        this.headers = options.headers;
+        this.request = options.request;
```

Comment:
I suspect we can continue further and inline the contents of `Request` into `Resource` as well.